[Nago] - Grid 보기, List 보기, 통합 구독, 페이지네이션 구현

Nago730 opened the pull request:
# ✅ 완료 작업 목록 1

## 🔹 개요

ListView 컴포넌트를 정적 UI에서 동적 데이터 기반 렌더링 방식으로 개선하고, 사용자 상호작용 기능을 추가했습니다.

## 🔹 구현 내용

- **JSON 데이터를 파싱하여 뉴스카드 동적 렌더링 구현**  
  - 하드코딩된 UI 컴포넌트들을 데이터 기반 렌더링 방식으로 변경  
  - NewsCard 컴포넌트에서 mainNews와 subNewsList 데이터 매핑 처리  

- **카테고리 클릭 인터랙션 기능 구현**  
  - 카테고리 선택 시 해당 카테고리의 뉴스 목록 필터링  
  - 선택된 카테고리 시각적 피드백 추가  

- **뉴스 링크 기능 구현**  
  - 메인 뉴스 제목 클릭 시 해당 기사로 이동  
  - 서브 뉴스 목록의 각 제목 클릭 시 해당 기사로 이동  
  - 새 탭에서 링크 열기 및 보안 속성 적용  

- **페이지 이동 버튼 컴포넌트 구현**  
  - NextPageBtn, PrevPageBtn 공통 컴포넌트 생성 및 스타일링  
  - disabled 상태에 따른 표시/숨김 처리  

- **ListView 페이지 이동 기능 개발**  
  - 페이지 인덱스 상태 관리 및 이동 핸들러 구현  
  - 페이지 범위에 따른 버튼 활성화/비활성화 적용  

- **페이지 이동 버튼 컴포넌트 통합 및 개선**  
  - GridView 전용 버튼을 Common 디렉토리의 공통 버튼으로 통합  
  - 컴포넌트 중복 제거 및 상태 처리 통일  

---

# ✅ 완료 작업 목록 2

## 🔹 개요

MediaContext를 생성하고 구독 상태 관리 및 데이터 전달 로직을 개선하였으며, 구독 버튼 컴포넌트를 통합하여 코드의 일관성과 재사용성을 높였습니다.

## 🔹 구현 내용

- **MediaContext 생성 및 구독 로직 추가**  
  - MediaContext를 생성하여 구독 여부와 구독 토글 로직을 중앙에서 관리  
  - GridView에 전달할 데이터를 MediaContext에서 관리하도록 개선 

- **페이지 상태 관리 개선**  
  - 전체 언론사와 구독한 언론사별로 현재 페이지(headerIndex, pageIndex)를 각각 독립적으로 기억하도록 상태 관리 로직 개선  

- **구독 버튼 컴포넌트 통합**  
  - 구독/해지 버튼을 하나의 공통 컴포넌트(SubscriptionButton)로 통합  
  - 배경색과 구독 상태에 따른 스타일 변화를 props로 제어  
  - hover 효과 및 transition을 일관성 있게 적용하여 UI 개선  

- **페이지 상태 관리 개선**
  - 전체 언론사와 구독한 언론사별로 현재 페이지(headerIndex, pageIndex)를 각각 독립적으로 기억하도록 상태 관리 로직 개선

---

# ✅ 완료 작업 목록 3

## 🔹 개요

모달(Alert) 상태를 전역으로 관리하기 위해 `AlertContext`를 생성하고, `SubscriptionButton`에서 모달 렌더링을 분리하였습니다. 이를 통해 모달의 위치 제약을 해소하고, 컴포넌트 간 책임 분리를 통해 재사용성과 유지보수성을 높였습니다.

## 🔹 구현 내용

- **AlertContext 생성 및 전역 상태 관리 도입**  
  - `show`, `pressId` 등의 상태를 포함한 `AlertContext` 생성  
  - 모달(Alert)을 전역 상태로 제어하여 원하는 위치에서 쉽게 렌더링 가능

- **Alert 컴포넌트 리팩토링**  
  - pressId만 받아 내부에서 pressName을 매핑하도록 단순화  
  - 구독 해지 로직과 Alert UI를 독립적으로 처리  

- **SubscriptionButton 리팩토링**  
  - Alert 관련 상태와 로직을 제거하고, context의 `openAlert(pressId)`만 호출하도록 변경  
  - 버튼 내부 로직이 간결해지고 재사용성이 향상됨  

- **모달 위치 지정 방식 개선**  
  - 모달에 `position: absolute` 적용 시 기준이 되는 부모(`relative`)를 특정 위치에서 설정 가능하도록 구조 변경  
  - 모달이 상위 컴포넌트에서 렌더링되므로 위치 제약 해소됨

---

# ✅ 완료 작업 목록 4

## 🔹 개요

Media 관련 전역 상태를 Context 기반으로 관리하도록 구조를 개선하고,  
GridView와 ListView 각각의 전용 상태를 Context로 분리하여 컴포넌트 간 의존성을 낮췄습니다.

## 🔹 구현 내용

- **MediaContext 분리 및 상태 이전**  
  - 기존 Media 컴포넌트 내 로컬 상태를 MediaContext로 이동  
  - 탭 상태(`tabType`), 뷰 타입(`viewType`), 구독 언론사 Map 등의 전역 상태 관리 구현  
  - 구독/구독 해지 핸들러 `handlePressSubscription` 함수 포함

- **GridContext 생성 및 상태 분리**  
  - GridView 전용 상태(`gridItemList`)를 GridContext로 이동  
  - 최초 로딩 시 Grid 데이터를 정제하고 셔플하는 로직을 `useEffect`로 처리  
  - GridProvider가 조건부 마운트 시마다 재생성되는 문제를 해결하기 위해 ListView에서도 공통으로 감싸는 방식으로 통일  

- **ListContext 생성 및 상태 이전**  
  - ListView 전용 상태(`headerIndexAll`, `headerIndexSubscribed`)를 ListContext로 분리  
  - 구독 해지 시 인덱스 범위 체크 및 자동 보정 로직 관리  

- **커스텀 훅 생성 및 적용**  
  - `useMediaContext`, `useGridContext`, `useListContext` 커스텀 훅 생성  
  - 각 뷰 컴포넌트에서 Context 직접 접근 대신 커스텀 훅 사용으로 가독성 및 유지보수성 향상

---
# 🛠 설계 과정에서의 주요 고민과 해결

## 🔹 문제 상황 및 해결 방안 1

- **문제 상황**  
GridView 컴포넌트의 크기가 커짐에 따라 컴포넌트 분리 여부를 고민했습니다. 현재 컴포넌트는 크게 4가지 역할을 수행합니다:
1. 데이터 준비 및 필터링 (상태 관리)
2. 페이지네이션 데이터 처리
3. 페이지 이동 핸들러
4. UI 렌더링

```javascript
function GridView({ pressData, tabType }: GridViewProps) {
  // 1. 데이터 준비 및 필터링
  const [gridItems] = useState(() => formatGridViewData(pressData))
  const { subscribedPressMap } = useContext(SubscriptionContext)

  const filteredPressList = useMemo(() => {
    if (tabType === 'subscribed') {
      return gridItems.filter(press => subscribedPressMap.has(press.pid))
    }
    return gridItems
  }, [gridItems, subscribedPressMap, tabType])

  // 2. 페이지네이션 데이터 처리
  const totalPages = Math.ceil(filteredPressList.length / MAX_GRID_PRESS_COUNT)
  const [pageIndex, setPageIndex] = useState(1)

  const currentPageItems = useMemo(() => {
    const startIndex = (pageIndex - 1) * MAX_GRID_PRESS_COUNT
    const items = [...filteredPressList.slice(startIndex, startIndex + MAX_GRID_PRESS_COUNT)]

    while (items.length < MAX_GRID_PRESS_COUNT) {
      items.push({ pid: `empty-${items.length}` } as PressDataType)
    }

    return items.map((item, idx) => (
      <Grid key={`grid-${item.pid}`} pressData={item} isEmpty={item.pid.startsWith('empty')} />
    ))
  }, [filteredPressList, pageIndex])

  // 3. 페이지 이동 핸들러
  const handleNextPage = () => {
    if (pageIndex < totalPages) {
      setPageIndex(prev => prev + 1)
    }
  }

  const handlePrevPage = () => {
    if (pageIndex > 1) {
      setPageIndex(prev => prev - 1)
    }
  }

  // 4. 렌더링
  return (
    <Container>
      {currentPageItems}
      <PrevPageBtn onClick={handlePrevPage} disabled={pageIndex === 1} />
      <NextPageBtn onClick={handleNextPage} disabled={pageIndex === totalPages} />
    </Container>
  )
}
```
- **해결 방안**  

컴포넌트를 현재 구조로 유지하기로 결정했습니다. 그 이유는:

1. **높은 응집도**
   - 각 부분이 긴밀하게 연결되어 데이터 흐름이 순차적이고 명확함
   - 상태(filteredPressList, pageIndex)가 여러 로직에서 공유됨

2. **단일 책임 원칙 준수**
   - 컴포넌트의 전체 목적이 "그리드 형태의 페이지네이션된 Press 데이터 표시"라는 하나의 책임에 집중
   - 억지로 분리 시 오히려 책임이 모호해질 수 있음

3. **불필요한 복잡성 방지**
   - 컴포넌트 분리 시 발생할 수 있는 문제들 회피:
     - props drilling
     - 상태 관리 복잡화
     - 컴포넌트 간 의존성 증가
     - 가독성 저하

결과적으로, 현재의 응집도 높은 단일 컴포넌트 구조가 유지보수와 가독성 측면에서 더 효율적이라고 판단했습니다.

---

## 🔹 문제 상황 및 해결 방안 2

- **문제 상황**  
  - `Media` 컴포넌트는 `GridView`와 `ListView` 두 가지 뷰 타입을 조건부 렌더링으로 전환함  
    ```tsx
    viewType === 'grid' ? (
      <GridView pressData={data} tabType={tabType} />
    ) : (
      <ListView pressData={data} tabType={tabType} />
    )
    ```
  - `GridView`는 매번 동일한 `pressData`를 props로 받지만, 최초 마운트 시 **랜덤하게 섞인 데이터를 기반으로 렌더링**해야 하는 요구사항이 존재  
  - 이를 위해 `useEffect` 내에서 의존성 배열을 빈 배열(`[]`)로 설정하여 최초 한 번만 데이터를 셔플하도록 구현했으나, 문제는 `viewType`이 변경되어 `GridView`가 언마운트될 때 발생  
  - `GridView`가 언마운트되면 내부 상태(`useState`, `useEffect`)가 모두 초기화되므로, 다시 `GridView`로 전환될 때마다 매번 셔플 로직이 재실행되는 문제가 있었음

- **해결 방안**  
  - **랜덤 셔플 데이터를 지속적으로 유지하기 위한 별도의 컨텍스트**를 도입  
  - `MediaContext`를 생성하고 `Media` 컴포넌트를 해당 컨텍스트로 감싸 상태를 중앙 집중식으로 관리  
  - `gridItems` 상태를 `MediaProvider` 상단에서 관리하고, `GridView`에서는 이 데이터를 참조만 하도록 리팩토링  
  - 최초 렌더링 시 `gridItems`가 비어있을 경우에만 셔플 로직이 실행되도록 조건 추가  

    ```tsx
    // MediaContext
    function MediaProvider({ children }: MediaProviderProps) {
      const [gridItems, setGridItems] = useState<PressDataType[]>([])
      // + 구독 상태 로직들
      return (
        <MediaContext.Provider value={{ gridItems, setGridItems, ... }}>
          {children}
        </MediaContext.Provider>
      )
    }

    // GridView 내부
    function GridView({ pressData, tabType }: GridViewProps) {
      const { gridItems, setGridItems } = useContext(MediaContext)

      useEffect(() => {
        if (gridItems.length === 0) {
          setGridItems(shuffle(getGridViewData(pressData)))
        }
      }, [])
    }
    ```

---

## 🔹 문제 상황 및 해결 방안 3

- **문제 상황**  
모달(Alert)을 `SubscriptionButton` 내부에서 렌더링했을 때, `absolute` 위치 기준이 버튼 기준으로 고정되어 UI 배치가 어려웠습니다. 또한, 알림창을 띄우기 위해 버튼 컴포넌트에 모달 로직까지 포함시켜야 해 재사용성과 유지보수성이 떨어졌습니다.

- **해결 방안**  
  - 전역 컨텍스트(`AlertContext`)를 생성하여 모달 상태와 로직을 전역에서 관리
  - `SubscriptionButton`에서는 오직 `pressId`만 넘겨주고, `openAlert(pressId)`만 호출하는 구조로 변경
  - 모달은 상위 컴포넌트에서 위치를 자유롭게 지정할 수 있도록 렌더링 위치를 이동  
  - 결과적으로 **SRP(단일 책임 원칙)**을 지키면서도 **UI 배치 문제**를 해결

Nago730 committed:
refactor: onViewChange -> setCurrentView로 변경

Nago730 committed:
feat: 언론사 구독 토글 기능 추가

- GridView에서 각 언론사 항목에 구독 토글 기능 구현
- 구독 상태 변경 시 구독한 언론사 수 동적으로 갱신
- 구독한 언론사 항목에 hover 시 UnSubscribeBtn 표시
- 구독 해제 기능 추가로 구독 목록에서 언론사 제거 가능

Nago730 committed:
refactor: 변수명 명확화
currentView -> viewType
setCurrentView -> setViewType
currentTab -> tabType
setCurrentTab -> setTabType

Nago730 committed:
feat: 뉴스 리스트뷰 동적 렌더링 구현

- JSON 데이터 기반 뉴스카드 동적 렌더링 구현
- 카테고리 클릭 기능 추가
- 메인뉴스 및 서브뉴스 링크 연결 기능 추가

이전: 하드코딩된 정적 UI
이후: JSON 데이터 기반 동적 렌더링

Nago730 committed:
feat: 페이지 이동 버튼 구현

- NextPageBtn, PrevPageBtn 컴포넌트 생성
- 버튼 클릭 시 페이지 이동 기능 구현
- disabled 상태에 따른 버튼 표시/숨김 처리
- ListView에 페이지 이동 버튼 적용

Nago730 committed:
refactor: 페이지 이동 버튼 컴포넌트 통합

- GridView 전용 버튼 컴포넌트 제거
- Common의 NextPageBtn, PrevPageBtn으로 통합
- GridView에서 공통 버튼 컴포넌트 적용

Nago730 committed:
refactor: 컴포넌트명 변경 MainLayout > Main

crongro commented:
> ```js
>       {currentPageItems}
> ```

응집도 차원에서 고심한 과정이 좋네요. 
전 이정도 복잡도면, 분리해도 좋다고 생각해요.
아마도 GridItemList 이걸 포함하는 컴포넌트를 분리하고, 관련 로직도 같이 보낼거 같아요.

지금도 이유도 있고 좋습니다.

crongro commented on src/components/Main/Media/GridView/index.tsx:
```diff
@@ -0,0 +1,74 @@
+const PAGE_GRID_NUMBER = 24
+
+function GridView({ pressData, setPressData }: GridViewProps) {
+  const totalPages = Math.ceil(pressData.length / 24) || 1
```
PAGE_GRID_NUMBER를 사용했으면 여기서도 사용하시지

crongro commented on src/components/Main/Media/GridView/index.tsx:
```diff
@@ -0,0 +1,74 @@
+    }
+  }
+
+  const handlePrevPage = () => {
+    if (currentPage > 1) {
+      setCurrentPage(prev => prev - 1)
+    }
+  }
```
전 이런건 그냥 한 줄로 표현하는것도 좋아보임(개.취)

const handlePrevPage = () => currentPage > 1 && setCurrentPage(prev => prev - 1)

crongro commented on src/components/Main/Media/ListView/NewsCard/index.tsx:
```diff
@@ -0,0 +1,54 @@
+export interface newsInfoType {
+  pid: string
+  name: string
+  regDate: string
```
date도 string이군요. 가급적 Date 사용할 수 있으면 더 좋을듯

crongro commented on src/components/Main/Media/index.tsx:
```diff
@@ -0,0 +1,67 @@
+}
+
+function Media() {
+  const { data: pressData, loading } = useFetch('/mock/press_by_category.json')
```
커스텀 훅 시도 좋네요. 유용하게 써보고 다듬어보세요.

crongro reviewed (approved):
고민하던 것 빼고는 컴포넌트 크기도 작고 명확하게 잘 분리했네요.

Nago730 committed:
feat: 구독 상태 관리를 위한 SubscriptionContext 추가

- 구독 상태를 저장하고 관리하기 위한 SubscriptionContext 생성
- Media 컴포넌트를 SubscriptionContext 제공자로 감쌈
- GridView와 ListView를 SubscriptionContext를 활용하도록 리팩토링
- tabType이 'subscribed'일 때 데이터를 필터링하고 렌더링하는 로직 구현
- GridView와 ListView 컴포넌트에 구독된 데이터를 위한 렌더링 로직 추가

Nago730 committed:
refactor: 컨텍스트 이름 변경
- SubscriptionContext > MediaContext

Nago730 committed:
refactor: state 이름 변경
- buttonIndex > headerIndex
- setButtonIndex > setHeaderIndex

Nago730 committed:
refactor: tabType별 독립적인 페이지 상태 관리 구현

- ListView와 GridView의 페이지 상태를 tabType('all', 'subscribed')별로 분리
- 각 탭의 페이지 상태가 독립적으로 유지되도록 개선
- 페이지 상태 관리 로직 단순화

Nago730 committed:
feat: 카테고리 간 자연스러운 페이지 전환 구현

- 첫 번째 뉴스카드에서 이전 버튼 클릭 시 이전 카테고리 마지막 뉴스카드로 이동
- 마지막 뉴스카드에서 다음 버튼 클릭 시 다음 카테고리 첫 번째 뉴스카드로 이동
- 카테고리 간 순환적 탐색 가능하도록 페이지네이션 로직 개선

Nago730 committed:
refactor: 구독 버튼 컴포넌트 통합

- 구독/해지 버튼을 하나의 공통 컴포넌트(SubscriptionButton)로 통합
- 배경색, 구독 상태에 따른 스타일 변화를 props로 제어
- hover 효과 및 transition 일관성 있게 적용
- ListView와 GridView에서 중복된 버튼 로직 제거

crongro commented on src/contexts/MediaContext.tsx:
```diff
@@ -0,0 +1,64 @@
+}
+
+const MediaContext = createContext<MediaContextType>({
+  subscribedPressMap: new Map(),
```
map을 썼네요. object 와의 차이점을 잘 비교해보세요.

Nago730 replied:
구독한 순서대로 언론사를 렌더링해야 하는 기능이 있는데, map은 삽입 순서를 보장합니다. map의 삽입 순서를 이용해서 언론사를 순서대로 렌더링할 수 있습니다!!

crongro replied:
좋아요. 구조적인 차이도 공부해보세요.

crongro commented on src/contexts/MediaContext.tsx:
```diff
@@ -0,0 +1,64 @@
+import { createContext, useState, ReactNode } from 'react'
+
+interface PressDataType {
```
Type이런거 뒤에 안붙여도 돼요. 
interface인데 type으로 착각할 수도 있고요.

crongro commented on src/contexts/MediaContext.tsx:
```diff
@@ -0,0 +1,64 @@
+  return (
+    <MediaContext.Provider
+      value={{
+        subscribedPressMap,
+        setSubscribedPressMap,
+        gridItems,
+        setGridItems,
+        handlePressSubscription,
```
하나의 컨텍스트에 너무 많은 것을 넣는 것도 주의가 필요해요.
불필요한 리렌더링이 너무 많이 발생할지 몰라서요.

crongro reviewed (approved):
ts 적용도 잘했고, context활용도 괜찮네요.

Nago730 committed:
refactor: 페이지네이션 개선
- prevBtn 클릭에 의해 headerIndex가 줄어들면 maxPageIndex로 설정정

Nago730 committed:
feat: 구독 해지 확인을 위한 알럿 컴포넌트 구현

- AlertContext 생성하여 알럿 상태 관리
- 구독 해지 확인을 위한 Alert 컴포넌트 구현
- 언론사 이름과 조사(을/를) 동적 표시 기능 추가

crongro commented on src/components/Common/SubscriptionBtn.tsx:
```diff
@@ -1,38 +1,51 @@
+  const handleClick = () => {
+    if (isSubscribed) {
+      setPressId(pressId)
+      setShowAlert(true)
```
setShowAlert 라고 하면 알럿으로 보여줄 메시지를 파라미터로 전달해야 할것 같아요.
지금든 얼럿창은 나오는데 내용은 뭘까? 그런 궁금증이 생길듯

crongro reviewed (approved):
그러고보니 이번에 타입스크립트까지 적용했군요.
수고하셨어요.

Nago730 committed:
feat: Media 상태 Context 분리 및 GridContext 생성

- Media 상태를 MediaContext로 분리하여 전역 관리
- Grid 전용 상태를 위한 GridContext 생성
- useMediaContext, useGridContext 커스텀 훅 구현 및 적용

Nago730 committed:
refacor: useMediaContext의 확장자명 tsx에서 ts로 변경경

Nago730 committed:
refactor: 초기 데이터 처리를 GridContext의 위임
- 기존 GridView에서 useEffect로 처리리

Nago730 committed:
feat: ListContext 생성

- List 상태를 ListContext로 분리하여 관리
- useListContext 커스텀 훅 구현 및 적용
